autodiff/forward/utils/derivative.hpp

```diff
@@ -0,0 +1,230 @@
+};
+
+/// The keyword used to denote the variables *with respect to* the derivative is calculated.
+template<typename... Args>
+auto wrt(Args&&... args)
+{
+    return Wrt<Args&&...>{ std::forward_as_tuple(std::forward<Args>(args)...) };
+}
+
+/// The keyword used to denote the variables *at* which the derivatives are calculated.
+template<typename... Args>
+auto at(Args&&... args)
+{
+    return At<Args&&...>{ std::forward_as_tuple(std::forward<Args>(args)...) };
+}
+
+/// The keyword used to denote the direction vector *along* which the derivatives are calculated.
+template<typename... Args>
+auto along(Args&&... args)
+{
+    return Along<Args&&...>{ std::forward_as_tuple(std::forward<Args>(args)...) };
+}
+
+/// Seed each dual number in the **wrt** list using its position as the derivative order to be seeded.
+/// Using `seed(wrt(x, y, z), 1)` will set the 1st order derivative of `x`, the
+/// 2nd order derivative of `y`, and the 3rd order derivative of `z` to 1. If
+/// these dual numbers have order greater than 3, then the last dual number will
+/// be used for the remaining higher-order derivatives. For example, if these
+/// numbers are 5th order, than the 4th and 5th order derivatives of `z` will be
+/// set to 1 as well. In this example, `wrt(x, y, z)` is equivalent to `wrt(x,
+/// y, z, z, z)`. This automatic seeding permits derivatives `fx`, `fxy`,
+/// `fxyz`, `fxyzz`, and `fxyzzz` to be computed in a more convenient way.
+template<typename Var, typename... Vars, typename T>
+auto seed(const Wrt<Var&, Vars&...>& wrt, T&& seedval)
+{
+    constexpr static auto N = Order<Var>;
+    constexpr static auto size = 1 + sizeof...(Vars);
+    static_assert(size <= N, "It is not possible to compute higher-order derivatives with order greater than that of the autodiff number (e.g., using wrt(x, x, y, z) will fail if the autodiff numbers in use have order below 4).");
+    For<N>([&](auto i) constexpr {
+        if constexpr (i < size)
+            seed<i.index + 1>(std::get<i>(wrt.args), seedval);
+        else
+            seed<i.index + 1>(std::get<size - 1>(wrt.args), seedval); // use the last variable in the wrt list as the variable for which the remaining higher-order derivatives are calculated (e.g., derivatives(f, wrt(x), at(x)) will produce [f0, fx, fxx, fxxx, fxxxx] when x is a 4th order dual number).
+    });
+}
+
+template<typename... Vars>
+auto seed(const Wrt<Vars...>& wrt)
+{
+    seed(wrt, 1.0);
+}
+
+template<typename... Vars>
+auto unseed(const Wrt<Vars...>& wrt)
+{
+    seed(wrt, 0.0);
+}
+
+template<typename... Args, typename... Vecs>
+auto seed(const At<Args...>& at, const Along<Vecs...>& along)
+{
+    static_assert(sizeof...(Args) == sizeof...(Vecs));
+
+    ForEach(at.args, along.args, [&](auto& arg, auto&& dir) constexpr {
+        if constexpr (isVector<decltype(arg)>) {
+            static_assert(isVector<decltype(dir)>);
+            assert(arg.size() == dir.size());
+            const size_t len = dir.size();
```

Comment:
For modern compilers, we don't need to create this `len` variable. It should optimize loop
```cpp
for(size_t i = 0; i < dir.size(); ++i)
```

Reply:
OK